### PR TITLE
target/ppc: Implemented vsx xvi*ger* instructions

### PR DESCRIPTION
**Pull request by alqotel**
Implemented integer GER instructions, both prefixed, unprefixed, with and without saturation.

### CODE REVIEW

File: target/ppc/int_helper.c [hunk 1]
```diff
@@ -910,6 +910,79 @@ VCT(uxs, cvtsduw, u32)
...
+        pmsk = (uint8_t)(mask >> 8) & 0xFF;                             \
+        opmsk = (mask >> 32) & 0xFF;                                    \
+        a = cpu_vsr_ptr(env, a_r);                                      \
+        b = cpu_vsr_ptr(env, b_r);                                      \
+        at = cpu_vsr_ptr(env, at_r);                                    \
+        sat = (bool)(opmsk & SAT_MASK);                                 \
+        for (i = 0; i < 4; i++) {                                       \
+            for (j = 0; j < 4; j++) {                                   \
+                if ((1 << (3-i) & xmsk) && (1 << (3-j) & ymsk)) {       \
+                    if (!(opmsk & AC_MASK)) {                           \
```
Comment (lffpires) [inline target/ppc/int_helper.c:940]: The order of these checks (acc/sat) doesn't look right.
You should first calculate psum (already including at[i].VsrW(j) when acc is set), and then chop/clamp (again, not tested :)):

```
        if ((xmsk & xmsk_bit) && (ymsk & ymsk_bit)) {
            psum = 0;
            <psum = prod0 + ... + prod<RANK-1>>
            if (acc) {
                psum += at[i].VsrW(j);
            }
            if (sat) {
                if (psum > (1LL << 31) - 1) {
                    psum = (1LL << 31) - 1;
                    set_vscr_sat(env);
                } else if (psum < -(1LL << 31)) {
                    psum = -(1LL << 31) ;
                    set_vscr_sat(env);
                }
            }
            at[i].VsrW(j) = (int32_t)psum;
```

File: target/ppc/int_helper.c [hunk 2]
```diff
@@ -910,6 +910,79 @@ VCT(uxs, cvtsduw, u32)
...
+                        at[i].VsrW(j) = 0;                              \
+                    }                                                   \
+                    psum = 0;                                           \
+                    for (k = 0; k < rank; k++) {                        \
+                        if (1 << (rank-1 -k) & pmsk) {                  \
+                            va = a->VsrS##el((rank*i + k)); \
+                            vb = (bits != 8) ?                          \
+                                  b->VsrS##el((rank*j + k)):      \
+                                  b->Vsr##el((rank*j + k));             \
```
[the remark this answers is not in the document]
Comment (mferst) [reply target/ppc/int_helper.c]: It is different in the ISA, `xvi8*` uses `EXTS` on `X` and `EXTZ` on `Y`.

File: target/ppc/int_helper.c [hunk 3]
```diff
@@ -910,6 +910,79 @@ VCT(uxs, cvtsduw, u32)
 VCT(sxs, cvtsdsw, s32)
 #undef VCT
 
+#define XVIGER_NAME(bits, rank)                                         \
+                    glue(glue(helper_xvi, bits), glue(ger, rank))
+
+#define VsrN(i) VsrB((i >> 1)) >> ((i+1) % 2)*4 & 0xF
```
[the remark this answers is not in the document]
Comment (mferst) [reply target/ppc/int_helper.c:916]: `VsrN` was a bad idea in the end. Because of the differences with `xvi8`, I suggested the changes in https://github.com/PPC64/qemu/pull/26/commits/ec8c278d293e4de2930e8cbb099510f9649599f2, and now we can drop `VsrN` and access `VsrB`/`VsrSB` directly in the `GER` macro. We still need a better name for this macro though, [GER](https://spec.oneapi.io/versions/latest/elements/oneMKL/source/domains/blas/ger.html) is the name of the entire operation.

File: target/ppc/int_helper.c [hunk 4]
```diff
@@ -910,6 +910,79 @@ VCT(uxs, cvtsduw, u32)
...
+#define VsrSN(i) VsrB((i >> 1)) >> ((i+1) % 2)*4 & 0xF | (0xF ^ (-1))
+
+#define EXTS(num, nb)                                                   \
+             (((((num) >> (nb - 1)) & 1) == 0) ?                        \
+             num :                                                      \
+             ((((1 << nb) - 1) ^ (-1)) | (num)))
+
+#define XVIGER(bits, rank, el)                                          \
+    void XVIGER_NAME(bits, rank) (CPUPPCState *env,                     \
```
[the remark this answers is not in the document]
Comment (mferst) [reply target/ppc/int_helper.c:919]: Adding an extra argument to these helpers seems a bit wasteful considering everything fits in 32-bits. Maybe we should take the `TCGMemOpIdx/make_memop_idx` route:
```c
typedef enum GEROp {
    /* Lower 16-bits packs XMSK, YMSK and PMSK */
    GER_SAT = 0x10000,    /* Saturate */
    GER_ACC = 0x20000,    /* Accumulate*/
    GER_NM  = 0x40000,    /* Negative multiply */
    GER_NA  = 0x80000,    /* Negative accumulate */
    /* Combinations of the above. */
    GER_PP = GER_ACC,
    GER_SPP = GER_SAT | GER_ACC,
    GER_NP = GER_ACC | GER_NM,
    GER_PN = GER_ACC | GER_NA,
    GER_NN = GER_ACC | GER_NM | GER_NA,
}

static inline GEROp make_gerop_mask(GEROp op, uint8_t xmask, uint8_t ymask,
                                    uint8_t pmask)
{
    xmask &= 0xF;
    ymask &= 0xF;
    pmask &= 0xFF;
    return op | (pmask << 8) | (ymask << 4) | xmask;
}

static inline uint8_t get_gerop_xmask(GEROp op)
{
    return op & 0xF;
}

static inline uint8_t get_gerop_ymask(GEROp op)
{
    return (op >> 4) & 0xF;
}

static inline uint8_t get_gerop_pmask(GEROp op)
{
    return (op >> 8) & 0xFF;
}
```

[the remark this answers is not in the document]
Comment (lffpires) [reply target/ppc/int_helper.c:919]: I don't believe having an extra parameter here is really a problem. Unless there's a significant performance impact, I would favor clarity.

To put things in perspective, we're being a lot more wasteful by using 3 i32 parameters to pass xa (6 bits), xb (6 bits) and at (3 bits). :)

File: target/ppc/insn32.decode [hunk 5]
```diff
@@ -52,6 +52,12 @@
 %xx2_xb         1:1 11:5
 @XX2            ...... ..... ... uim:2 ..... ......... .. &XX2 xt=%xx2_xt xb=%xx2_xb
 
+&XX3            ax bx tx
```
[the remark this answers is not in the document]
Comment (alqotel) [reply target/ppc/insn32.decode]: Ok, changed it.

File: target/ppc/insn32.decode [hunk 6]
```diff
@@ -52,6 +52,12 @@
...
+%xx3_ax         2:1 16:5
+%xx3_bx         1:1 11:5
+%xx3_tx         23:3 !function=times_4
```
[the remark this answers is not in the document]
Comment (alqotel) [reply target/ppc/insn32.decode]: Moved the multiplication by 4 to vsx-impl.c.inc

File: target/ppc/insn32.decode [hunk 7]
```diff
@@ -52,6 +52,12 @@
...
+@XX3            ...... ... -- ..... ..... ........ ..-  &XX3 ax=%xx3_ax bx=%xx3_bx tx=%xx3_tx
```
[the remark this answers is not in the document]
Comment (alqotel) [reply target/ppc/insn32.decode]: Good point, changed it

File: target/ppc/int_helper.c [hunk 8]
```diff
@@ -910,6 +910,79 @@ VCT(uxs, cvtsduw, u32)
...
+        ymsk = (uint8_t)(mask >> 4) & 0x0F;                             \
+        pmsk = (uint8_t)(mask >> 8) & 0xFF;                             \
+        opmsk = (mask >> 32) & 0xFF;                                    \
+        a = cpu_vsr_ptr(env, a_r);                                      \
+        b = cpu_vsr_ptr(env, b_r);                                      \
+        at = cpu_vsr_ptr(env, at_r);                                    \
+        sat = (bool)(opmsk & SAT_MASK);                                 \
+        for (i = 0; i < 4; i++) {                                       \
+            for (j = 0; j < 4; j++) {                                   \
+                if ((1 << (3-i) & xmsk) && (1 << (3-j) & ymsk)) {       \
```
[the remark this answers is not in the document]
Comment (alqotel) [reply target/ppc/int_helper.c]: Changed it, but @mferst suggested maybe moving the xmsk "if" to outside the second "for", because if that bit is 0 then the entire line should be set to 0

[the remark this answers is not in the document]
Comment (lffpires) [reply target/ppc/int_helper.c]: Good point!

File: target/ppc/insn64.decode [hunk 9]
```diff
@@ -94,23 +94,23 @@ XXSPLTI32DX     000001 01 0000 -- -- ................ \
 ## Vector GER instruction
 
 PMXVI4GER8      000001 11 1001 -- - - pmsk:8 xmsk:4 ymsk:4              \
```
Comment (lffpires) [inline target/ppc/insn64.decode:96]: Sorry for not spotting this in the first pass, but maybe we should move the field definitions of xmsk and ymsk, which never change in MMIRR:XX3, to @MMIRR_XX3. And, in the instructions, keep only pmsk, which varies in size.

File: target/ppc/int_helper.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (lffpires) [reply target/ppc/int_helper.c]: When you skip the whole line, don't forget to set all of the elements in that line to zero

File: target/ppc/translate/vsx-impl.c.inc [hunk 10]
```diff
@@ -2124,44 +2124,60 @@ static void gen_xvxsigdp(DisasContext *ctx)
...
+ */
+#define XVIGER(name, func_name, saturate, accumulate)                   \
+    static bool trans_##name (DisasContext *ctx, arg_XX3 *a)            \
     {                                                                   \
-        uint64_t mask;                                                  \
+        uint32_t mask, op_flag;                                         \
         REQUIRE_INSNS_FLAGS2(ctx, ISA310);                              \
+        if (unlikely((a->xa >= a->at*4 && a->xa <= a->at*4 + 3) ||      \
+                     (a->xb >= a->at*4 && a->xb <= a->at*4 + 3))) {     \
+            return false;                                               \
```
Comment (lffpires) [inline target/ppc/translate/vsx-impl.c.inc:2140]: Since you already know the form is invalid in the trans() function, instead of returning false, you should:
```
gen_invalid(ctx);
return true;
```
(generate code for the invalid form exception and return true to indicate the decode itself was successful)

File: target/ppc/translate/vsx-impl.c.inc [hunk 11]
```diff
@@ -2124,44 +2124,60 @@ static void gen_xvxsigdp(DisasContext *ctx)
...
+        }                                                               \
                                                                         \
-        mask = (uint64_t)0xFFFFFFFF;                                    \
-        mask |= ((uint64_t)op_mask) << 32;                              \
-        gen_helper_##func_name (cpu_env, tcg_constant_i32(a->ax),       \
-                                         tcg_constant_i32(a->bx),       \
-                                         tcg_constant_i32(a->tx),       \
-                                         tcg_constant_i64(mask));       \
+        mask = (uint32_t)0xFFFFFFFF;                                    \
+        op_flag = ((saturate << 1) | accumulate) << 2;                  \
```
Comment (lffpires) [inline target/ppc/translate/vsx-impl.c.inc:2144]: Having nested shifts it is harder to know which bits are being set (``saturate`` is being shifted 1 bit, and then 2 more bits).
IMO, it would be simpler to read:
```
op_flag = (saturate << 3) | (accumulate << 2);
```

File: target/ppc/translate/vsx-impl.c.inc [hunk 12]
```diff
@@ -2124,44 +2124,60 @@ static void gen_xvxsigdp(DisasContext *ctx)
     tcg_temp_free_i64(xbl);
 }
 
-#define XVIGER(name, func_name, op_mask)                                \
-    static bool trans_##name (DisasContext *ctx, arg_##name * a)        \
+/*
```
[the remark this answers is not in the document]
Comment (lffpires) [reply target/ppc/translate/vsx-impl.c.inc:2127]: Is it correct to say that the ger functions can be as follows?
- either with or without saturation
- either with or without multiply+accumulate (there's no ger instruction JUST with multiply or JUST with accumulate - it has to be both)
- if it has multiply+accumulate, there can be 4 different scenarios:
  - negative multiply, negative accumulate
  - negative multiply, positive accumulate
  - positive multiply, negative accumulate
  - positive multiply, positive accumulate

If so, I would document these bits as:
```
Bit 3:
  0: Without saturate
  1: With saturate
Bits 2-0:
  0xx: No multiply, no accumulate
  100: Negative multiply, negative accumulate
  101: Negative multiply, positive accumulate
  110: Positive multiply, negative accumulate
  111: Positive multiply, positive accumulate
```
(regardless of using or not using an enum. But, after seeing these different combinations, I think using an enum would help)

File: target/ppc/int_helper.c [hunk 13]
```diff
@@ -963,33 +964,39 @@ VCT(sxs, cvtsdsw, s32)
     }
 
 
-#define GER(bits, rank, el)                                             \
+#define GER_MULT                                                        \
     do {                                                                \
-        va = a->VsrN((rank*i + k));                                     \
+        va = (a->VsrB((i*8 + k)/2) >> ((k+1) % 2)*4) & 0xF;             \
```
Comment (lffpires) [inline target/ppc/int_helper.c:969]: I still think it would be clearer, even using GER_MULT, to use the GET_Vsr* macros to distinguish between N/B/H - but then it would be a single GER_MULT, with varying GET_Vsr*.

I understand this what you're trying to do is to have a single XVIGER that can be reused by all of these instructions and also all of the floating point ger instructions as well?

To decide, it would be good for us (and @mferst) to review what the XVIGER code would look like, with all the scenarios handled. What do you think?

BTW, for floating point, will we also need to redefine psum (not as int64_t)?

File: target/ppc/translate/vsx-impl.c.inc [hunk 14]
```diff
@@ -2124,44 +2124,60 @@ static void gen_xvxsigdp(DisasContext *ctx)
...
+        op_flag = ((saturate << 1) | accumulate) << 2;                  \
+        gen_helper_##func_name (cpu_env, tcg_constant_i32(a->xa),       \
+                                         tcg_constant_i32(a->xb),       \
+                                         tcg_constant_i32(a->at * 4),   \
+                                         tcg_constant_i32(mask),        \
+                                         tcg_constant_i32(op_flag));    \
         return true;                                                    \
     }                                                                   \
-    static bool trans_PM##name (DisasContext *ctx, arg_PM##name * a)    \
+    static bool trans_PM##name (DisasContext *ctx, arg_MMIRR_XX3 *a)    \
```
Comment (lffpires) [inline target/ppc/translate/vsx-impl.c.inc:2152]: Same comments as the other trans() function

**Review by lffpires — COMMENTED** [review]
A few more comments.

And I'm not sure if I mentioned it in this PR or another, but you're missing spaces around operators (+, -, <<, etc.)

File: target/ppc/cpu.h [hunk 15]
```diff
@@ -2672,4 +2677,21 @@ void dump_mmu(CPUPPCState *env);
...
+    GER_NA  = 0x1,    /* Negative accumulate */
+    /* Combinations of the above. */
+    GER_NOP = 0,
+    GER_PP  = GER_ACC,
+    GER_SPP = GER_SAT | GER_ACC,
+    GER_NP  = GER_ACC | GER_NM,
+    GER_PN  = GER_ACC | GER_NA,
+    GER_NN  = GER_ACC | GER_NM | GER_NA,
+};
 #endif /* PPC_CPU_H */
```
Comment (alqotel) [inline target/ppc/cpu.h:2697]: I've put here this enum, but I don't think this is the best place, so does anyone have any suggestions?

File: target/ppc/int_helper.c [hunk 16]
```diff
@@ -910,6 +910,102 @@ VCT(uxs, cvtsduw, u32)
...
+                    for (k = 0, pmsk_bit = 1 << (RANK-1); k < RANK;     \
+                         k++, pmsk_bit >>= 1) {                         \
+                        if (pmsk_bit & pmsk) {                          \
+                            GER_MULT(BITS, RANK, EL);                   \
+                        }                                               \
+                    }                                                   \
+                    if (acc) {                                          \
+                        psum += at->VsrW(j);                            \
+                    }                                                   \
+                    if (sat && (psum > (1LL<<31) - 1)) {                \
```
[the remark this answers is not in the document]
Comment (alqotel) [reply target/ppc/int_helper.c:974]: I've been thinking of using just INT_MIN and INT_MAX here with a cast if necessary to improve readability, what do you think?

File: target/ppc/int_helper.c [hunk 17]
```diff
@@ -910,6 +910,102 @@ VCT(uxs, cvtsduw, u32)
...
+                        set_vscr_sat(env);                              \
+                        at->VsrSW(j) = (1LL<<31) - 1;                   \
+                    } else if (sat && psum < -(1LL<<31)) {              \
+                        set_vscr_sat(env);                              \
+                        at->VsrSW(j) = -(1LL<<31);                      \
+                    } else {                                            \
+                        at->VsrSW(j) = (int32_t) psum;                  \
+                    }                                                   \
+                } else {                                                \
+                    at->VsrW(j) = 0;                                    \
```
[the remark this answers is not in the document]
Comment (alqotel) [reply target/ppc/int_helper.c:984]: I used ```VsrW(j)``` because the ISA defines that if not ```XMSK.bit[i] & YMSK.bit[j]``` then ```ACC[AT][i][j] ← 0x0000_0000```, so I thought that since it's attributing an hex, then unsigned seemed more fitting.

File: target/ppc/translate/vsx-impl.c.inc [hunk 18]
```diff
@@ -2124,6 +2124,55 @@ static void gen_xvxsigdp(DisasContext *ctx)
...
+    static bool trans_##name (DisasContext *ctx, arg_XX3 *a)            \
+    {                                                                   \
+        uint32_t mask;                                         \
+        REQUIRE_INSNS_FLAGS2(ctx, ISA310);                              \
+        if (unlikely((a->xa/4 == a->at) || (a->xb/4 == a->at))) {       \
+            gen_invalid(ctx);                                           \
+            return true;                                                \
+        }                                                               \
+                                                                        \
+        mask = (uint32_t)0xFFFFFFFF;                                    \
```
[the remark this answers is not in the document]
Comment (alqotel) [reply target/ppc/translate/vsx-impl.c.inc:2137]: Yes, looking back it was unnecessary

File: target/ppc/translate/vsx-impl.c.inc [hunk 19]
```diff
@@ -2124,6 +2124,55 @@ static void gen_xvxsigdp(DisasContext *ctx)
     tcg_temp_free_i64(xbl);
 }
 
+#define XVIGER(name, func_name, op_flag)                   \
+    static bool trans_##name (DisasContext *ctx, arg_XX3 *a)            \
+    {                                                                   \
+        uint32_t mask;                                         \
+        REQUIRE_INSNS_FLAGS2(ctx, ISA310);                              \
+        if (unlikely((a->xa/4 == a->at) || (a->xb/4 == a->at))) {       \
```
[the remark this answers is not in the document]
Comment (alqotel) [reply target/ppc/translate/vsx-impl.c.inc:2132]: Ok, also this macro is not really needed, I'll just use a ```do_ger``` function with ```TRANS``` and ```TRANS64``` macros

File: target/ppc/int_helper.c (hunk 16 above)
[the remark this answers is not in the document]
Comment (lffpires) [reply target/ppc/int_helper.c:974]: Although I've never seen INT_MIN/INT_MAX values other than 2^31-1/-2^31, I believe they are platform-dependent, and could assume different values. In this case, you need specifically 2^31-1 and -2^31.

That being said, I believe you can safely use INT32_MIN and INT32_MAX.

File: target/ppc/int_helper.c (hunk 17 above)
[the remark this answers is not in the document]
Comment (lffpires) [reply target/ppc/int_helper.c:984]: I would use VsrSW, just to avoid confusion. 0x0000_0000 is equally 'signed 0', and both options (assigning to VsrW and to VsrSW) will perform the exact same thing.

File: target/ppc/translate/vsx-impl.c.inc (hunk 19 above)
[the remark this answers is not in the document]
Comment (lffpires) [reply target/ppc/translate/vsx-impl.c.inc:2132]: ok

File: target/ppc/int_helper.c [hunk 20]
```diff
@@ -910,6 +910,23 @@ VCT(uxs, cvtsduw, u32)
 VCT(sxs, cvtsdsw, s32)
 #undef VCT
 
+/*
```
Comment (lffpires) [inline target/ppc/int_helper.c:913]: I don't think we should define these functions. Testing directly against the enum options should be simple enough.

File: target/ppc/translate/vsx-impl.c.inc [hunk 21]
```diff
@@ -2124,19 +2124,33 @@ static void gen_xvxsigdp(DisasContext *ctx)
...
 }
 
+/*
+ * Packed VSX Integer GER Flags
+ * 00 - no accumulation no saturation
+ * 01 - accumulate but no saturation
+ * 10 - no accumulation but with saturation
+ * 11 - accumulate with saturation
+ */
+static int pack_flags_xvi(int acc, int sat)
```
Comment (lffpires) [inline target/ppc/translate/vsx-impl.c.inc:2134]: We can also revert these changes and go back to using the enum options directly.

File: target/ppc/translate/vsx-impl.c.inc [hunk 22]
```diff
@@ -2156,28 +2170,28 @@ static bool do_ger_XX3(DisasContext *ctx, arg_XX3 *a, int op_flag,
...
-TRANS64(PMXVI8GER4PP  , do_ger_MMIRR_XX3, GER_PP , gen_helper_xvi8ger4)
-TRANS64(PMXVI8GER4SPP , do_ger_MMIRR_XX3, GER_SPP, gen_helper_xvi8ger4)
-TRANS64(PMXVI16GER2   , do_ger_MMIRR_XX3, GER_NOP, gen_helper_xvi16ger2)
-TRANS64(PMXVI16GER2PP , do_ger_MMIRR_XX3, GER_PP , gen_helper_xvi16ger2)
-TRANS64(PMXVI16GER2S  , do_ger_MMIRR_XX3, GER_SAT, gen_helper_xvi16ger2)
-TRANS64(PMXVI16GER2SPP, do_ger_MMIRR_XX3, GER_SPP, gen_helper_xvi16ger2)
+    return do_ger_MMIRR_XX3(ctx, &m, acc, sat, helper);
+}
+
+TRANS(XVI4GER8    , do_ger_XX3, 0, 0, gen_helper_xvi4ger8)
```
Comment (lffpires) [inline target/ppc/translate/vsx-impl.c.inc:2180]: Here too we can undo the changes and use the enum options (and also in do_ger_XX3).

File: target/ppc/translate/vsx-impl.c.inc [hunk 23]
```diff
@@ -2151,14 +2156,13 @@ static bool do_ger_MMIRR_XX3(DisasContext *ctx, arg_MMIRR_XX3 *a, int acc, int s
...
 
-    mask = a->pmsk << 8 | a->ymsk << 4 | a->xmsk;
+    mask = pack_masks(a->pmsk, a->ymsk, a->xmsk);
     helper (cpu_env, tcg_constant_i32(a->xa), tcg_constant_i32(a->xb),
             tcg_constant_i32(a->at * 4), tcg_constant_i32(mask),
             tcg_constant_i32(op_flag));
     return true;
 
 }
-
```
Comment (lffpires) [inline target/ppc/translate/vsx-impl.c.inc]: Missing line?

File: target/ppc/int_helper.c [hunk 24]
```diff
@@ -927,6 +927,21 @@ static inline bool get_acc(uint32_t flags)
     return flags & 0x1;
 }
 
+static inline uint8_t get_xmsk(uint32_t packed_masks)
```
Comment (lffpires) [inline target/ppc/int_helper.c:930]: Move these 3 functions to internal.h? And also prefix them with 'ger_', since they're 100% specific to the ger instructions? (ger_get_xmsk, etc.)

File: target/ppc/translate/vsx-impl.c.inc [hunk 25]
```diff
@@ -2136,6 +2136,11 @@ static int pack_flags_xvi(int acc, int sat)
     return (sat << 1) | acc;
 }
 
+static int pack_masks(int pmsk, int ymsk, int xmsk)
```
Comment (lffpires) [inline target/ppc/translate/vsx-impl.c.inc:2139]: Move this function to internal.h? And also prefix it with 'ger_', since they're 100% specific to the ger instructions?

File: target/ppc/translate/vsx-impl.c.inc [hunk 26]
```diff
@@ -2136,6 +2136,11 @@ static int pack_flags_xvi(int acc, int sat)
...
+{
```
Comment (lffpires) [inline target/ppc/translate/vsx-impl.c.inc:2140]: Suggested comment to add before the group of ger_get_*msk and ger_pack_masks functions:
```
/*
 * Auxiliary functions to pack/unpack masks for GER instructions.
 *
 * Packed format:
 *  Bits 0-3: xmsk
 *  Bits 4-7: ymsk
 *  Bits 8-15: pmsk
 */
```

File: target/ppc/translate/vsx-impl.c.inc [hunk 27]
```diff
@@ -2136,6 +2136,11 @@ static int pack_flags_xvi(int acc, int sat)
...
+    return (pmsk << 8) | (ymsk << 4) | (xmsk);
```
Comment (lffpires) [inline target/ppc/translate/vsx-impl.c.inc:2141]: It would be safer to mask pmsk, ymsk and xmsk to the appropriate sizes too:
```
return (pmsk & 0xFF) << 8 | (ymsk & 0xF) << 4 | (xmsk & 0xF);
```
